src/Data/Text/Internal/Reverse.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{-# OPTIONS_HADDOCK not-home #-}
+
+-- | Implements 'reverse', using efficient C routines by default.
+module Data.Text.Internal.Reverse (reverse) where
+
+#if !defined(PURE_HASKELL)
+import GHC.Exts as Exts
+import Control.Monad.ST.Unsafe (unsafeIOToST)
+import Foreign.C.Types (CSize(..))
+#else
+import Control.Monad.ST (ST)
+import Data.Text.Internal.Encoding.Utf8 (utf8LengthByLeader)
+#endif
+import Prelude hiding (reverse)
+import Data.Text.Internal (Text(..))
+import Control.Monad.ST (runST)
+import qualified Data.Text.Array as A
+
+-- | /O(n)/ Reverse the characters of a string.
+--
+-- Example:
+--
+-- $setup
+-- >>> T.reverse "desrever"
+-- "reversed"
+reverse ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  Text -> Text
+#if defined(PURE_HASKELL)
+reverse (Text ba off len) = runST $ do
+    dest <- A.new len
+    _ <- reversePoints ba off dest (len - 1)
+    result <- A.unsafeFreeze dest
+    pure $ Text result 0 len
+
+reversePoints
+    :: A.Array -- ^ Input array
+    -> Int -- ^ Input index
+    -> A.MArray s -- ^ Output array
+    -> Int -- ^ Output index
+    -> ST s ()
+reversePoints _ _ _ p_out | p_out < 0 = pure ()
+reversePoints ba p_in dest p_out =
+    let pointLength = utf8LengthByLeader (A.unsafeIndex ba p_in)
+    in do
+        A.copyI pointLength dest (p_out - pointLength + 1) ba p_in
+        reversePoints ba (p_in + pointLength) dest (p_out - pointLength)
```

Comment:
I didn't benchmark it, but a general rule is that passing unchanged arguments through recursive calls is suboptimal for performance. A usual pattern is 
```haskell
reversePoints ba x dest y = go x y 
  where 
    go pIn pOut = .... go pInt' pOut'
```

Reply:
I tried this, and all the microbenchmarks surprisingly fared a lot worse.

```patch
diff --git a/src/Data/Text/Internal/Reverse.hs b/src/Data/Text/Internal/Reverse.hs
index 153611f..3c0f95c 100644
--- a/src/Data/Text/Internal/Reverse.hs
+++ b/src/Data/Text/Internal/Reverse.hs
@@ -46,12 +46,13 @@ reversePoints
     -> A.MArray s -- ^ Output array
     -> Int -- ^ Output index
     -> ST s ()
-reversePoints _ _ _ p_out | p_out < 0 = pure ()
-reversePoints ba p_in dest p_out =
-    let pointLength = utf8LengthByLeader (A.unsafeIndex ba p_in)
-    in do
-        A.copyI pointLength dest (p_out - pointLength + 1) ba p_in
-        reversePoints ba (p_in + pointLength) dest (p_out - pointLength)
+reversePoints src x dest y = go x y where
+    go _ pOut | pOut < 0 = pure ()
+    go pIn pOut =
+        let pLen = utf8LengthByLeader (A.unsafeIndex src pIn)
+        in do
+            A.copyI pLen dest (pOut - pLen + 1) src pIn
+            go (pIn + pLen) (pOut - pLen)
 #else
 reverse (Text (A.ByteArray ba) off len) = runST $ do
     marr@(A.MutableByteArray mba) <- A.new len
```

```
Benchmark text-benchmarks: RUNNING...
All
  Pure
    tiny
      reverse
        Text:     OK (0.28s)
          63.8 ns ± 5.3 ns, 79% more than baseline
        LazyText: OK (0.22s)
          50.2 ns ± 2.8 ns, 10% more than baseline
    ascii-small
      reverse
        Text:     OK (0.19s)
          732  μs ±  54 μs, 101% more than baseline
        LazyText: OK (0.12s)
          472  μs ±  43 μs, 29% more than baseline
    ascii
      reverse
        Text:     OK (2.02s)
          637  ms ±  19 ms, 110% more than baseline
        LazyText: OK (1.26s)
          395  ms ± 5.5 ms, 32% more than baseline
    english
      reverse
        Text:     OK (0.33s)
          41.9 ms ± 1.5 ms, 98% more than baseline
        LazyText: OK (0.19s)
          26.4 ms ± 1.5 ms, 32% more than baseline
    russian
      reverse
        Text:     OK (0.32s)
          75.4 μs ± 6.3 μs, 83% more than baseline
        LazyText: OK (0.21s)
          51.4 μs ± 3.0 μs, 25% more than baseline
    japanese
      reverse
        Text:     OK (0.21s)
          49.4 μs ± 2.8 μs, 103% more than baseline
        LazyText: OK (0.29s)
          34.6 μs ± 2.9 μs, 42% more than baseline

All 12 tests passed (6.42s)
Benchmark text-benchmarks: FINISH
```

Reply:
Surprisingly, I can reproduce your measurements, but, looking at Core, I'm fairly certain that this is some sort of benchmarking quirk. You can compare yourself by putting
```haskell
{-# OPTIONS_GHC -ddump-to-file -ddump-simpl -dsuppress-all -dno-suppress-type-signatures #-}
```

* Let's go with `go pIn pOut`, Core output is much nicer. 
* You most certainly want `go !_ pOut`, not just `go _ pOut`, otherwise GHC believes that this function is lazy in `pIn` and would not unbox it. 
* Common subexpression elimination pass is not powerful enough to share `pOut - pLen` between two expressions, please do it manually.
* Another microoptimisation is to pass `reversePoints ba off dest len`, check `pOut <= 0`, and then you shave off `+1` in `A.copyI pLen dest (pOut - pLen) src pIn`.